sycl/doc/extensions/experimental/sycl_ext_oneapi_spec_constant_length_alloca.asciidoc

```diff
@@ -0,0 +1,232 @@
+users can make use of dynamic memory allocations, e.g., `std::vector`, but this
+is not the case in SYCL device code.
+
+SYCL specialization constants (SYCL Section 4.9.5.) can be used to represent
+constants whose values can be set dynamically during the execution of a SYCL
+application, but that will not change when a SYCL kernel function is
+invoked. This way, specialization constants could be used to implement SYCL
+private arrays whose size is given during the execution of the SYCL
+application. There is no possible way of implementing this using `std::array`,
+as the size of such containers must be known at compile time, so we propose to
+define a new `private_alloca` function whose size is specified using SYCL
+specialization constants.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_SPEC_CONSTANT_LENGTH_ALLOCA` to one of the values defined
+in the table below. Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New Aspect for Specification Constant-Length Allocations
+
+This extension adds a new device aspect:
+
+[source,c++]
+----
+namespace sycl {
+enum class aspect : /*unspecified*/ {
+  ext_oneapi_scla
+};
+} // namespace sycl
+----
+
+The `ext_oneapi_scla` aspect indicates that the device is capable of using the
+`private_alloca` API defined in the following sections.
+
+=== The SCLA API
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+template <typename ElementType, auto &SpecName,
+          access::decorated DecorateAddress>
+private_ptr<ElementType, DecorateAddress>
+private_alloca(kernel_handler &kh);
+
+template <typename ElementType, std::size_t Alignment, auto &SpecName,
+          access::decorated DecorateAddress>
+private_ptr<ElementType, DecorateAddress>
+aligned_private_alloca(kernel_handler &kh);
+} // namespace sycl::ext::oneapi::experimental
+----
+
+_Mandates_: `ElementType` must be a cv-unqualified trivial type and `SpecName`
+must be a reference to a specialization constant of integral `value_type`. In
+the case of `aligned_private_alloca`, `Alignment` must be a power of 2
+fundamental alignment stricter than the alignment requirement of `ElementType`.
+
+_Effects_: `h.get_specialization_constant<size>()` elements of type
+`ElementType` are allocated and default initialized in private memory.
+
+_Returns_: A pointer to a default initialized region of private memory of
+`h.get_specialization_constant<size>()` elements of type
```

Comment:
Checking that I understand this correctly ... you're saying here that the returned pointer always points to a buffer if "size" elements, where "size" is the value of the specialization constant.  Correct?  That must mean that every call to `private_alloca` return the same pointer value.  For example, every iteration of the loop below returns the same value for `ptr`:

```
constexpr specialization_id<int> size(1);

void run(queue q, const float *in, float *out, size_t n) {
  q.submit([&](handler &h) {
    h.set_specialization_constant<size>(n);
    h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler kh) {
      for (int i = 0;  i < N;  i++) {
        auto ptr = private_alloca<float, size, access::decorated::yes>(kh);
        // All iterations get the same value for "ptr"
    });
  });
}
```

Do I understand correctly?

If so, I think the name "alloca" is confusing because it implies that each call returns a separate allocation.  If my understanding is correct, it seems like this functions is pretty similar to [sycl_ext_oneapi_work_group_specific](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_specific.asciidoc), except that it returns memory in the private address space rather than in the local address space.

I have some more comments to make, but I'll stop here to check that my understanding is correct.

Reply:
You're right to some extent. `private_alloca` would be something in between, as we can have:
```c++
constexpr specialization_id<int> size(1);

void run(queue q, const float *in, float *out, size_t n) {
  q.submit([&](handler &h) {
    h.set_specialization_constant<size>(n);
    h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler kh) {
      auto ptr0 = private_alloca<float, size, access::decorated::yes>(kh);
      auto ptr1 = private_alloca<float, size, access::decorated::yes>(kh);
      // ptr0 and ptr1 point to different allocations
    });
  });
}
```

`private_alloca` effectively declares an array. At runtime, after specialization constants are instantiated, your code would be equivalent to
```c++
constexpr specialization_id<int> size(1);

void run(queue q, const float *in, float *out, size_t n) {
  q.submit([&](handler &h) {
    h.set_specialization_constant<size>(n);
    h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler kh) {
      for (int i = 0;  i < N;  i++) {
        float ptr[/*constant n value*/];
        auto ptr = private_alloca<float, size, access::decorated::yes>(kh);
        // All iterations get the same value for "ptr"
    });
  });
}
```

Maybe `alloca` is not the best word to describe these semantics...

Reply:
Do you think something like:

```c++
constexpr specialization_id<int> size(1);

void run(queue q, const float *in, float *out, size_t n) {
  q.submit([&](handler &h) {
    h.set_specialization_constant<size>(n);
    h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler kh) {
      for (int i = 0;  i < N;  i++) {
        auto ptr = private_array<float, size, access::decorated::yes>(kh);
        // All iterations get the same value for "ptr"
    });
  });
}
```
i.e., renaming to `private_array` would make this somewhat clearer?

Reply:
Maybe I'm missing something, but it seems like it would be much clearer to expose this as a class, rather than a free function.  For example:

```
constexpr specialization_id<int> size(1);

void run(queue q, const float *in, float *out, size_t n) {
  q.submit([&](handler &h) {
    h.set_specialization_constant<size>(n);
    h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler kh) {
      work_item_array<float, size> array_a{kh};
      work_item_array<float, size> array_b{kh};

      // array_a and array_b have different storage
      array_a[0] = 3.14f;
      array_b[0] = 2.718f;
    });
  });
}
```

This seems a lot clearer because each variable of type `work_item_array` has it's own unique storage.  With the free function, there are confusing rules about when the storage is unique.  Two dynamic calls to the same static instance of the function return the same storage, while calls to different static instances of the function return unique storage.

I also see it as an advantage that the syntax above is similar to the [sycl_ext_oneapi_work_group_specific](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_specific.asciidoc) extension.  It's easier for people to learn SYCL if we use the same pattern in multiple places, rather then using different patterns everywhere.

Is there some problem I'm missing with the `work_item_array` pattern I show above?

Reply:
> Maybe I'm missing something, but it seems like it would be much clearer to expose this as a class,

We also looked into options to have that as a class (including a similar syntax to `sycl_ext_oneapi_work_group_specific`. From the proposal:

> The big design constraint stems from the unknown allocation size at compile
time. C++ does not support variable length arrays and complete type sizes must
be known at compile time. Thus, the free function interface returning a pointer
to private memory is the better way to represent this construct in C++. Lifetime
of the underlying memory region was a concern too, but the current design with
automatic storage duration for the allocated memory region closely follows what
the user would get from a stack-allocated array.

SCLA as a dynamically sized type was ruled out then. A more similar design to the extension you mention was discarded also, due to restrictions in private memory allocations in underlying APIs. Effectively, the biggest difference between SYCL private memory and other kind of memories is that it is allocated in the call stack conceptually in underlying backends, e.g., in SPIR-V, it corresponds to [_Function Storage Class_](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_storage_class):

> Visible only within the declaring function of the current invocation.

Users writing something like:

```c++
work_item_specific<int[]> foo() {
  return {};
}
```

Where the returned pointer is deallocated as soon as `foo` returns, would run into some surprises seeing RAII violated.

Summing up: Private memory specifics made dynamic memory allocations really tricky in SYCL. We chose not to have this as a type and either violate C++ type rules and have a dynamically sized type or violate RAII. The alternative we found was the free function design, documenting that the pointer points to an automatic storage array.

I hope this throws some light in the design process we went through and why we ended up with this (imperfect) design.

Reply:
We met offline.  Marking this conversation as resolved.